op-challenger/runner/runner.go

```diff
@@ -0,0 +1,244 @@
+}
+
+func NewRunner(logger log.Logger, cfg *config.Config) *Runner {
+	return &Runner{
+		log: logger,
+		cfg: cfg,
+		m:   NewMetrics(),
+	}
+}
+
+func (r *Runner) Start(ctx context.Context) error {
+	if !r.running.CompareAndSwap(false, true) {
+		return errors.New("already started")
+	}
+	ctx, cancel := context.WithCancel(ctx)
+	r.ctx = ctx
+	r.cancel = cancel
+	if err := r.initMetricsServer(&r.cfg.MetricsConfig); err != nil {
+		return fmt.Errorf("failed to start metrics: %w", err)
+	}
+
+	rollupClient, err := dial.DialRollupClientWithTimeout(ctx, 1*time.Minute, r.log, r.cfg.RollupRpc)
+	if err != nil {
+		return fmt.Errorf("failed to dial rollup client: %w", err)
+	}
+
+	l1Client, err := dial.DialRPCClientWithTimeout(ctx, 1*time.Minute, r.log, r.cfg.L1EthRpc)
+	if err != nil {
+		return fmt.Errorf("failed to dial l1 client: %w", err)
+	}
+	caller := batching.NewMultiCaller(l1Client, batching.DefaultBatchSize)
+
+	for _, traceType := range r.cfg.TraceTypes {
+		r.wg.Add(1)
+		go r.loop(ctx, traceType, rollupClient, caller)
+	}
+
+	r.log.Info("Runners started")
+	return nil
+}
+
+func (r *Runner) loop(ctx context.Context, traceType types.TraceType, client *sources.RollupClient, caller *batching.MultiCaller) {
+	defer r.wg.Done()
+	t := time.NewTicker(1 * time.Minute)
+	defer t.Stop()
+	for {
+		if err := r.runOnce(ctx, traceType, client, caller); errors.Is(err, ErrUnexpectedStatusCode) {
+			r.log.Error("Incorrect status code", "type", traceType, "err", err)
+			r.m.RecordInvalid(traceType)
+		} else if err != nil {
+			r.log.Error("Failed to run", "type", traceType, "err", err)
+			r.m.RecordFailure(traceType)
+		} else {
+			r.log.Info("Successfully verified output root", "type", traceType)
+			r.m.RecordSuccess(traceType)
+		}
+		select {
+		case <-t.C:
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func (r *Runner) runOnce(ctx context.Context, traceType types.TraceType, client *sources.RollupClient, caller *batching.MultiCaller) error {
+	prestateHash, err := r.getPrestateHash(ctx, traceType, caller)
+	if err != nil {
+		return err
+	}
+
+	localInputs, err := r.createGameInputs(ctx, client)
+	if err != nil {
+		return err
+	}
+	dir, err := r.prepDatadir(traceType)
+	if err != nil {
+		return err
+	}
+	prestateSource := prestates.NewPrestateSource(
+		r.cfg.CannonAbsolutePreStateBaseURL,
+		r.cfg.CannonAbsolutePreState,
+		filepath.Join(dir, "prestates"))
+	logger := r.log.New("l1", localInputs.L1Head, "l2", localInputs.L2Head, "l2Block", localInputs.L2BlockNumber, "claim", localInputs.L2Claim, "type", traceType)
+	provider, err := createTraceProvider(logger, r.m, r.cfg, prestateSource, prestateHash, traceType, localInputs, dir)
+	if err != nil {
+		return fmt.Errorf("failed to create trace provider: %w", err)
+	}
+	hash, err := provider.Get(ctx, types.RootPosition)
+	if err != nil {
+		return fmt.Errorf("failed to execute trace provider: %w", err)
+	}
+	if hash[0] != mipsevm.VMStatusValid {
+		r.m.RecordInvalid(traceType)
```

Comment:
no need to record here. It's already handled by `loop` for the invalid error case (L103)